File: rnn.R
timestamp <- 100
# 皆代表特徵空間維度
input_features <- 32
output_features <- 64

random_array <- function(dim) {
  # prod(dim) 是維度中所有元素的乘積，這裡為32*100，runif用來創建均勻分布的隨機偏差
  array(runif(prod(dim)), dim = dim)
  # 根據指定的維度（dim）轉換成相應維度的數組
}

inputs <- random_array(dim = c(timestamp, input_features)) # 輸入數據: 這個例子為隨機噪聲
state_t <- rep_len(0, length = c(output_features)) # 初始狀態: 全零向量

# 創建隨機權重矩陣
W <- random_array(dim = c(output_features, input_features))
U <- random_array(dim = c(output_features, output_features))
b <- random_array(dim = c(output_features, 1))

output_sequence <- array(0, dim = c(timestamp, output_features))

for (i in 1 : nrow(inputs)) {
  
  input_t <- inputs[i,]
  
  # 維度：(64, 32) * (32, 1) + (64, 64) * (64, 1) + (64,1) = (64, 1)
  output_t <- tanh(as.numeric((W %*% input_t) + (U %*% state_t) + b)) # 將輸入與當前狀態(前一個輸出)結合
  output_sequence[i,] <- as.numeric(output_t)
  # 1:64
  state_t <- output_t # 更新下一個時間步的狀態
  
  cat('state :', state_t, "\n")
}

File: NN/src/rnn.rs
use ndarray::Array;
use rand::Rng;
// use crate::activation_function::tanh;

fn random_array(dim: usize) -> Array<f64, ndarray::IxDyn> {
    let mut rng = rand::thread_rng();
    Array::from_shape_fn(ndarray::IxDyn(&[dim]), |_| rng.gen_range(0.0..1.0))
}

pub fn recurrent() {

    let timestamp = 10;
    let input_features = 3;
    let output_features = 4;

    let inputs = random_array(timestamp * input_features).into_shape((timestamp, input_features)).unwrap();
    // (4, 1)
    let mut state_t = Array::zeros((output_features, 1));
    println!("inputs:{:?} state_t:{:?}", inputs.shape(), state_t.shape());

    let w = random_array(output_features * input_features).into_shape((output_features, input_features)).unwrap();
    let u = random_array(output_features * output_features).into_shape((output_features, output_features)).unwrap();
    // (4, 1)
    let b = random_array(output_features).into_shape((output_features, 1)).unwrap();
    println!("W:{:?} U:{:?} b:{:?}", w.shape(), u.shape(), b.shape());
    // (10, 4)
    let mut output_sequence = Array::zeros((timestamp, output_features));
    println!("output_sequence: {:?}", output_sequence.shape());

    for i in 0..timestamp {

        // let input_t = inputs.row(i);
        let input_t = inputs.row(i).insert_axis(ndarray::Axis(1));

        // (4, 3) * (3, 1) + (4, 4) * (4, 1) = (4, 1)
        let output_t = w.dot(&input_t) + u.dot(&state_t) + &b;
        println!("{:?} * {:?} + {:?} * {:?} + {:?}", w.shape() , input_t.shape(), u.shape(), state_t.shape(), b.shape());
        let output_t = output_t.mapv(|x| x.tanh());
        // 獲取第i個時間的輸出並保存output_t到output_sequence中
        output_sequence.row_mut(i).assign(&output_t);
        // println!("output_sequence: {:?}", output_sequence.row_mut(i));
        state_t = output_t.clone();
        
        println!("state: {:?}", state_t);
    }
}
